File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int i=0, j=0;
        int n=s.length();
        int longest=0;
        HashMap<Character, Integer> hMap=new HashMap<>();
        while(i<n){
            Character currChar=s.charAt(i);
            if(hMap.containsKey(currChar) && hMap.get(currChar)>=j){
                j=hMap.get(currChar)+1;
            }
            hMap.put(currChar, i);
            longest=Math.max(longest, i-j+1);
            i+=1;
        }
        return longest;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hMap = {}
        max_length = 0
        l = 0
        for r in range(0, len(s)):
            if s[r] in hMap and hMap[s[r]] >= l:
                l = hMap[s[r]] + 1
            hMap[s[r]] = r
            max_length = max(max_length, r - l + 1)
        return max_length

    # def lengthOfLongestSubstring(self, s: str) -> int:
    #     hset=set()
    #     l,length=0,0

    #     for r in range(len(s)):
    #         while s[r] in hset:
    #             hset.remove(s[l])
    #             l+=1

    #         hset.add(s[r])
    #         length=max(length,r-l+1)

    #     return length
